documentation/ha.markdown

```diff
@@ -0,0 +1,265 @@
+been copied to the replica. If the primary PuppetDB becomes unreachable
+in this state, exported resource queries will be redirected to the replica,
+which does not yet have the latest data.
+
+This problem is significantly mitigated by using the command_broadcast flag in
+puppetdb.conf, as recommended above. By pushing all data to the replica PuppetDB
+when doing the initial write, the replica should nearly always be up to date,
+but some failure cases still exist (for instance, intermittent connectivity to a
+replica followed by immediately losing the primary). Because of this, we don't
+recommend using PuppetDB HA in conjunction with exported resources for sensitive
+applications where the data is changing often. For typical applications, such as
+adding newly provisioned nodes to a load balancer's pool or registering them
+with a monitoring system, small state regressions will have very low impact.
+
+Q: *What durability guarantees does PuppetDB HA offer?*
+
+A: In the recommended configuration described above, there are corner cases
+which allow data loss. If you want to be absolutely certain that all data is
+stored on at least 2 (or more) PuppetDB servers, you can put this in your
+puppetdb.conf:
+
+        min_successful_submissions: 2
+
+With this configuration, a Puppet run will fail if any data cannot be written to
+at least 2 PuppetDB servers. This may be useful if you have 3 PuppetDB servers,
+for example.
+
+Operations
+-----
+
+PuppetDB provides several facilities to help you monitor the state of
+replication:
+
+## Structured Logging
+
+HA-related events are written to a log named ":sync". You can configure the
+handling of these events to fit your requirements.
+
+If you have configured structured logging as described in the
+[logging][Logging Configuration Guide], you will see additional attributes on each JSON log message.
+
+### Common fields
+
+* `phase`: The sync is divided into nested phases: `"sync"`, `"entity"`, `"record"`, and
+  `"deactivate"`. These are described in more detail below.
+
+* `event`: All sync log messages have an event field, indicating its position
+  within the phase. This is one of `"start"`, `"finished"`, or `"error"`.
+
+* `ok`:
+  * `finished` messages have the JSON boolean value `true`.
+  * `error` messages have the JSON boolean value `false`.
+
+* `elapsed`: `finished` and `error` messages have an `elapsed` field whose value
+  is the time span since the corresponding start event in milliseconds
+  (formatted as a JSON number).
+
+
+### Sync phase
+
+`sync` events have the following fields:
+
+* remote: The URL on the remote system from which data is being pulled.
+
+### Entity phase
+
+`entity` events surround the syncing for each type of record, i.e. catalogs,
+facts, or reports.
+
+* `entity`: The entity being processed. One of `"catalogs"`, `"facts"`, `"reports"`, or
+  `"nodes"`. Nodes is used only to sync node deactivation status.
```

Comment:
It should render ok. The double quotes are there because these are actually json strings.